compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -733,6 +859,33 @@ abstract class GenJSCode extends plugins.PluginComponent
           newBody)(methodDef.optimizerHints, None)(methodDef.pos)
     }
 
+    /** Moves all statements after the super constructor call.
+     *
+     *  This is used for the primary constructor of a Scala.js-defined JS
+     *  class, because those cannot access `this` before the super constructor
+     *  call.
+     *
+     *  scalac inserts statements before the super constructor call for early
+     *  initializers and param accessor initializers (including val's and var's
+     *  declared in the params). We move those after the super constructor
+     *  call, and are therefore executed later than for a Scala class.
+     */
+    private def moveAllStatementsAfterSuperConstructorCall(
+        body: js.Tree): js.Tree = {
+      val bodyStats = body match {
+        case js.Block(stats) => stats
+        case _               => body :: Nil
+      }
+
+      val (beforeSuper, superCall :: afterSuper) =
+        bodyStats.span(!_.isInstanceOf[js.JSSuperConstructorCall])
+
+      js.Block(
+          superCall ::
+          beforeSuper :::
+          afterSuper)(body.pos)
+    }
+
```

Comment:
Ahem... This method seems not very safe... While I can see how this is necessary, shouldn't we be a bit more conservative? Maybe detect patterns that we allow to move (and document) and if we find something else fail?

We can do this all in asserts, so it can be turned off if it is hindering for performance...